espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/adapter/espresso/AndroidMotionEvent.java

```diff
@@ -1,78 +1,89 @@
 package io.appium.espressoserver.lib.helpers.w3c.adapter.espresso;
 
-import android.os.SystemClock;
 import android.support.test.espresso.UiController;
+import android.view.MotionEvent;
 
+import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 
 import io.appium.espressoserver.lib.handlers.exceptions.AppiumException;
 import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.PointerType;
 import io.appium.espressoserver.lib.helpers.w3c.state.KeyInputState;
 
-import static android.view.MotionEvent.ACTION_DOWN;
+import static android.view.MotionEvent.ACTION_CANCEL;
 import static android.view.MotionEvent.ACTION_MOVE;
-import static android.view.MotionEvent.ACTION_UP;
 import static io.appium.espressoserver.lib.helpers.w3c.adapter.espresso.Helpers.extractButton;
 
 public class AndroidMotionEvent {
 
+    private static AndroidMotionEvent touchMotionEvent;
     private long downTime;
     private final UiController uiController;
     private static final Map<String, AndroidMotionEvent> motionEvents = new ConcurrentHashMap<>();
 
     private AndroidMotionEvent(UiController uiController) {
-        this.downTime = SystemClock.uptimeMillis();
         this.uiController = uiController;
-
-        // TODO: How do we distinguish STYLUS, MOUSE and TOUCH?
     }
 
-    public void pointerDown(long x, long y,
-                            Integer button, PointerType pointerType,
-                            final KeyInputState globalKeyInputState)
+    public MotionEvent pointerUpOrDown(List<Long> x, List<Long> y,
+                                       int action,
+                                       Integer button, PointerType pointerType,
+                                       final KeyInputState globalKeyInputState,
+                                       final MotionEvent downEvent,
+                                       final long eventTime)
             throws AppiumException {
 
         // TODO: Use globalKeyInputState to get metaState
         int metaState = 0;
 
-        (new MotionEventBuilder(uiController))
-                .setAction(ACTION_DOWN)
+        this.downTime = downEvent != null ? downEvent.getDownTime() : eventTime;
+
+        return (new MotionEventBuilder(uiController))
+                .setAction(action)
```

Comment:
in java world people usually use `with` prefix for builder methods instead of `set`